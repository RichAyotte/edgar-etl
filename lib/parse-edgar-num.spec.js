/**
 * @overview    CSV Parser stream spec
 * @author      Richard Ayotte
 * @copyright   Copyright Â© 2019 Richard Ayotte
 * @license     MIT License
 */

'use strict'

const fs = require('fs')
const {Writable} = require('stream')
const parseEdgarNum = require('./parse-edgar-num')

const getParsedNumbers = () => new Promise((resolve, reject) => {
	const parsedNumbers = []
	fs.createReadStream(`./sample-data/num-min.txt`)
	.pipe(parseEdgarNum)
	.pipe(new Writable({
		objectMode: true
		, write(data, _, done) {
			parsedNumbers.push(data)
			done()
		}
	}))
	.on('finish', () => {
		resolve(parsedNumbers)
	})
	.on('error', error => {
		reject(error)
	})
	.on('skip', error => {
		reject(error)
	})
})

describe('EDGAR number parser', () => {
	it(`Parses data into objects`, async () => {
		expect.assertions(1)
		const parsedNumbers = await getParsedNumbers()
		expect(parsedNumbers).toMatchSnapshot()
	})
})
