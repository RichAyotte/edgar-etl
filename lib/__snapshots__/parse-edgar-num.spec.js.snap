// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EDGAR number parser Parses data into objects 1`] = `
Array [
  Object {
    "adsh": "0001193125-09-186163",
    "coreg": "",
    "ddate": 20080630,
    "footnote": "",
    "qtrs": 1,
    "tag": "EarningsPerShareBasic",
    "uom": "USD",
    "value": 0.56,
    "version": "us-gaap/2008",
  },
  Object {
    "adsh": "0001193125-09-166661",
    "coreg": "",
    "ddate": 20080630,
    "footnote": "",
    "qtrs": 2,
    "tag": "StockIssuedOrVestedDuringPeriodValueStockAwards",
    "uom": "USD",
    "value": 30000000,
    "version": "0001193125-09-166661",
  },
  Object {
    "adsh": "0001193125-09-153203",
    "coreg": "",
    "ddate": 20090630,
    "footnote": "",
    "qtrs": 2,
    "tag": "AmortizationOfIntangibleAssets",
    "uom": "USD",
    "value": 54000000,
    "version": "us-gaap/2008",
  },
  Object {
    "adsh": "0000950123-09-024681",
    "coreg": "",
    "ddate": 20080630,
    "footnote": "",
    "qtrs": 2,
    "tag": "EffectOfExchangeRateOnCashAndCashEquivalents",
    "uom": "USD",
    "value": 1600000,
    "version": "us-gaap/2009",
  },
  Object {
    "adsh": "0001193125-09-171667",
    "coreg": "",
    "ddate": 20090630,
    "footnote": "",
    "qtrs": 3,
    "tag": "IncreaseDecreaseInCustomerAdvances",
    "uom": "USD",
    "value": 45054000,
    "version": "us-gaap/2008",
  },
  Object {
    "adsh": "0000060667-09-000096",
    "coreg": "",
    "ddate": 20090731,
    "footnote": "",
    "qtrs": 0,
    "tag": "AdditionalPaidInCapitalCommonStock",
    "uom": "USD",
    "value": 367000000,
    "version": "us-gaap/2009",
  },
  Object {
    "adsh": "0001104659-09-047977",
    "coreg": "",
    "ddate": 20080630,
    "footnote": "",
    "qtrs": 2,
    "tag": "EarningsPerShareDiluted",
    "uom": "USD",
    "value": 0.96,
    "version": "us-gaap/2009",
  },
  Object {
    "adsh": "0000950123-09-025125",
    "coreg": "",
    "ddate": 20090630,
    "footnote": "",
    "qtrs": 1,
    "tag": "NetIncomeLoss",
    "uom": "USD",
    "value": 113200000,
    "version": "us-gaap/2009",
  },
  Object {
    "adsh": "0000950123-09-030860",
    "coreg": "",
    "ddate": 20081231,
    "footnote": "",
    "qtrs": 0,
    "tag": "Liabilities",
    "uom": "USD",
    "value": 74048000000,
    "version": "us-gaap/2009",
  },
  Object {
    "adsh": "0000024741-09-000027",
    "coreg": "",
    "ddate": 20080630,
    "footnote": "",
    "qtrs": 2,
    "tag": "NetCashProvidedByUsedInFinancingActivities",
    "uom": "USD",
    "value": -206000000,
    "version": "us-gaap/2009",
  },
  Object {
    "adsh": "0000918160-09-000029",
    "coreg": "",
    "ddate": 20080630,
    "footnote": "",
    "qtrs": 2,
    "tag": "OtherOperatingIncome",
    "uom": "USD",
    "value": 9200000,
    "version": "us-gaap/2009",
  },
  Object {
    "adsh": "0001193125-09-155628",
    "coreg": "",
    "ddate": 20080630,
    "footnote": "",
    "qtrs": 2,
    "tag": "ProceedsFromRepaymentsOfOtherLongTermDebt",
    "uom": "USD",
    "value": 7000000,
    "version": "us-gaap/2008",
  },
  Object {
    "adsh": "0001193125-09-166495",
    "coreg": "",
    "ddate": 20090630,
    "footnote": "",
    "qtrs": 1,
    "tag": "InvestmentIncomeAndOther",
    "uom": "USD",
    "value": 26279000,
    "version": "0001193125-09-166495",
  },
  Object {
    "adsh": "0000950123-09-027072",
    "coreg": "",
    "ddate": 20080630,
    "footnote": "",
    "qtrs": 2,
    "tag": "GainLossOnDispositionOfAssets",
    "uom": "USD",
    "value": 0,
    "version": "us-gaap/2008",
  },
  Object {
    "adsh": "0000950123-09-033506",
    "coreg": "",
    "ddate": 20090731,
    "footnote": "",
    "qtrs": 0,
    "tag": "EntityCommonStockSharesOutstanding",
    "uom": "shares",
    "value": 111450654,
    "version": "dei/2009",
  },
  Object {
    "adsh": "0001193125-09-167378",
    "coreg": "",
    "ddate": 20090630,
    "footnote": "",
    "qtrs": 0,
    "tag": "CommonStockSharesAuthorized",
    "uom": "shares",
    "value": 1000000000,
    "version": "us-gaap/2008",
  },
  Object {
    "adsh": "0000899689-09-000034",
    "coreg": "",
    "ddate": 20081231,
    "footnote": "",
    "qtrs": 0,
    "tag": "DeferredCompensationLiabilityCurrentAndNoncurrent",
    "uom": "USD",
    "value": 69945000,
    "version": "us-gaap/2009",
  },
  Object {
    "adsh": "0000950123-09-031596",
    "coreg": "",
    "ddate": 20081231,
    "footnote": "",
    "qtrs": 0,
    "tag": "SelfInsuranceReserve",
    "uom": "USD",
    "value": 234500000,
    "version": "us-gaap/2009",
  },
  Object {
    "adsh": "0001047469-09-007370",
    "coreg": "",
    "ddate": 20081231,
    "footnote": "",
    "qtrs": 0,
    "tag": "OtherAssetsCurrent",
    "uom": "USD",
    "value": 235000000,
    "version": "us-gaap/2009",
  },
  Object {
    "adsh": "0001193125-09-162284",
    "coreg": "",
    "ddate": 20090630,
    "footnote": "",
    "qtrs": 2,
    "tag": "NetCashProvidedByUsedInFinancingActivitiesContinuingOperations",
    "uom": "USD",
    "value": -400000000,
    "version": "us-gaap/2009",
  },
  Object {
    "adsh": "0001104659-09-046313",
    "coreg": "",
    "ddate": 20080630,
    "footnote": "",
    "qtrs": 2,
    "tag": "OtherComprehensiveIncomeDerivativesQualifyingAsHedgesNetOfTaxPeriodIncreaseDecrease",
    "uom": "USD",
    "value": -4718000,
    "version": "us-gaap/2008",
  },
  Object {
    "adsh": "0001193125-09-163543",
    "coreg": "ParentCompany",
    "ddate": 20090630,
    "footnote": "",
    "qtrs": 1,
    "tag": "DividendsCommonStock",
    "uom": "USD",
    "value": 162000000,
    "version": "us-gaap/2008",
  },
  Object {
    "adsh": "0001193125-09-167155",
    "coreg": "",
    "ddate": 20080630,
    "footnote": "",
    "qtrs": 2,
    "tag": "IncomeLossFromContinuingOperationsPerBasicShare",
    "uom": "USD",
    "value": 0.23,
    "version": "us-gaap/2008",
  },
  Object {
    "adsh": "0001193125-09-154174",
    "coreg": "",
    "ddate": 20090630,
    "footnote": "",
    "qtrs": 4,
    "tag": "IncreaseDecreaseInAccountsReceivable",
    "uom": "USD",
    "value": 149000000,
    "version": "us-gaap/2008",
  },
  Object {
    "adsh": "0000950123-09-032660",
    "coreg": "",
    "ddate": 20090630,
    "footnote": "",
    "qtrs": 1,
    "tag": "IncomeLossFromDiscontinuedOperationsNetOfTaxPerBasicShare",
    "uom": "USD",
    "value": 0.01,
    "version": "us-gaap/2009",
  },
  Object {
    "adsh": "0001274494-09-000019",
    "coreg": "",
    "ddate": 20080630,
    "footnote": "",
    "qtrs": 2,
    "tag": "PaymentsToAcquireBusinessesNetOfCashAcquired",
    "uom": "USD",
    "value": 0,
    "version": "us-gaap/2009",
  },
  Object {
    "adsh": "0001193125-09-168917",
    "coreg": "",
    "ddate": 20081231,
    "footnote": "",
    "qtrs": 0,
    "tag": "OtherLiabilitiesNoncurrent",
    "uom": "USD",
    "value": 165500000,
    "version": "us-gaap/2008",
  },
  Object {
    "adsh": "0000950123-09-032642",
    "coreg": "",
    "ddate": 20090630,
    "footnote": "",
    "qtrs": 2,
    "tag": "InterestAndFeeIncomeLoansAndLeasesHeldForSale",
    "uom": "USD",
    "value": 21000000,
    "version": "us-gaap/2009",
  },
  Object {
    "adsh": "0001047469-09-007384",
    "coreg": "",
    "ddate": 20090630,
    "footnote": "",
    "qtrs": 2,
    "tag": "EarningsPerShareDiluted",
    "uom": "USD",
    "value": -0.58,
    "version": "us-gaap/2009",
  },
  Object {
    "adsh": "0001193125-09-167427",
    "coreg": "",
    "ddate": 20090630,
    "footnote": "",
    "qtrs": 0,
    "tag": "OtherLiabilitiesNoncurrent",
    "uom": "USD",
    "value": 3547000000,
    "version": "us-gaap/2009",
  },
  Object {
    "adsh": "0001193125-09-170623",
    "coreg": "",
    "ddate": 20090630,
    "footnote": "",
    "qtrs": 1,
    "tag": "NetOtherThanTemporaryImpairments",
    "uom": "USD",
    "value": 39728000,
    "version": "0001193125-09-170623",
  },
  Object {
    "adsh": "0000950123-09-027630",
    "coreg": "",
    "ddate": 20081231,
    "footnote": "",
    "qtrs": 0,
    "tag": "CommonStockSharesAuthorized",
    "uom": "shares",
    "value": 400000000,
    "version": "us-gaap/2009",
  },
  Object {
    "adsh": "0000040545-09-000049",
    "coreg": "SubsidiariesGECS",
    "ddate": 20090630,
    "footnote": "",
    "qtrs": 1,
    "tag": "NetIncomeLossAvailableToCommonStockholdersBasic",
    "uom": "USD",
    "value": 156000000,
    "version": "us-gaap/2009",
  },
  Object {
    "adsh": "0001193125-09-154888",
    "coreg": "",
    "ddate": 20081231,
    "footnote": "",
    "qtrs": 0,
    "tag": "CommonStockSharesAuthorized",
    "uom": "shares",
    "value": 2000000000,
    "version": "us-gaap/2008",
  },
  Object {
    "adsh": "0001193125-09-159733",
    "coreg": "",
    "ddate": 20090630,
    "footnote": "",
    "qtrs": 0,
    "tag": "PrepaidExpenseCurrent",
    "uom": "USD",
    "value": 284647000,
    "version": "us-gaap/2008",
  },
  Object {
    "adsh": "0000950123-09-026820",
    "coreg": "",
    "ddate": 20090630,
    "footnote": "",
    "qtrs": 0,
    "tag": "LongTermDebtNoncurrent",
    "uom": "USD",
    "value": 998562000,
    "version": "us-gaap/2009",
  },
  Object {
    "adsh": "0001047469-09-007382",
    "coreg": "",
    "ddate": 20080630,
    "footnote": "",
    "qtrs": 2,
    "tag": "CashAndCashEquivalentsPeriodIncreaseDecrease",
    "uom": "USD",
    "value": -12961000,
    "version": "us-gaap/2009",
  },
  Object {
    "adsh": "0000040545-09-000049",
    "coreg": "SubsidiariesGECS",
    "ddate": 20080630,
    "footnote": "",
    "qtrs": 2,
    "tag": "NetCashProvidedByUsedInFinancingActivities",
    "uom": "USD",
    "value": 33232000000,
    "version": "us-gaap/2009",
  },
  Object {
    "adsh": "0001193125-09-166658",
    "coreg": "",
    "ddate": 20080630,
    "footnote": "",
    "qtrs": 2,
    "tag": "PaymentsForRepurchaseOfCommonStock",
    "uom": "USD",
    "value": 2798400000,
    "version": "us-gaap/2008",
  },
  Object {
    "adsh": "0001193125-09-159733",
    "coreg": "",
    "ddate": 20090630,
    "footnote": "",
    "qtrs": 2,
    "tag": "ProceedsFromSaleAndMaturityOfAvailableForSaleSecurities",
    "uom": "USD",
    "value": 59009000,
    "version": "us-gaap/2008",
  },
  Object {
    "adsh": "0001145443-09-001878",
    "coreg": "",
    "ddate": 20090630,
    "footnote": "",
    "qtrs": 1,
    "tag": "OtherComprehensiveIncomeForeignCurrencyTransactionAndTranslationAdjustmentNetOfTaxPeriodIncreaseDecrease",
    "uom": "USD",
    "value": 15641000,
    "version": "us-gaap/2009",
  },
  Object {
    "adsh": "0000950123-09-027011",
    "coreg": "",
    "ddate": 20090630,
    "footnote": "",
    "qtrs": 2,
    "tag": "EarningsPerShareDiluted",
    "uom": "USD",
    "value": 1.22,
    "version": "us-gaap/2009",
  },
  Object {
    "adsh": "0000930413-09-003976",
    "coreg": "",
    "ddate": 20080630,
    "footnote": "",
    "qtrs": 2,
    "tag": "IncreaseDecreaseInRestrictedCash",
    "uom": "USD",
    "value": 0,
    "version": "us-gaap/2009",
  },
  Object {
    "adsh": "0000950123-09-025033",
    "coreg": "",
    "ddate": 20090630,
    "footnote": "",
    "qtrs": 2,
    "tag": "StockIssuedDuringPeriodValueStockDividend",
    "uom": "USD",
    "value": -93033000,
    "version": "us-gaap/2008",
  },
  Object {
    "adsh": "0000950123-09-028388",
    "coreg": "",
    "ddate": 20090630,
    "footnote": "",
    "qtrs": 0,
    "tag": "CommonStockValue",
    "uom": "USD",
    "value": 4647000,
    "version": "us-gaap/2009",
  },
  Object {
    "adsh": "0000909281-09-000056",
    "coreg": "",
    "ddate": 20090630,
    "footnote": "",
    "qtrs": 2,
    "tag": "NetIncomeLossAllocatedToLimitedPartners",
    "uom": "USD",
    "value": 151429000,
    "version": "us-gaap/2009",
  },
  Object {
    "adsh": "0000950123-09-031993",
    "coreg": "",
    "ddate": 20081231,
    "footnote": "",
    "qtrs": 0,
    "tag": "RetainedEarningsAccumulatedDeficit",
    "uom": "USD",
    "value": 2196235000,
    "version": "us-gaap/2008",
  },
  Object {
    "adsh": "0001047469-09-007382",
    "coreg": "",
    "ddate": 20090630,
    "footnote": "",
    "qtrs": 2,
    "tag": "NetCashProvidedByUsedInContinuingOperations",
    "uom": "USD",
    "value": -78152000,
    "version": "0001047469-09-007382",
  },
  Object {
    "adsh": "0001193125-09-179839",
    "coreg": "",
    "ddate": 20070630,
    "footnote": "",
    "qtrs": 4,
    "tag": "StockIssuedDuringPeriodValueEmployeeStockPurchasePlan",
    "uom": "USD",
    "value": 621000,
    "version": "us-gaap/2008",
  },
  Object {
    "adsh": "0001140361-09-017883",
    "coreg": "",
    "ddate": 20090630,
    "footnote": "",
    "qtrs": 2,
    "tag": "DeferredIncomeTaxExpenseBenefit",
    "uom": "USD",
    "value": -5400000,
    "version": "us-gaap/2009",
  },
  Object {
    "adsh": "0000950123-09-026820",
    "coreg": "",
    "ddate": 20090630,
    "footnote": "",
    "qtrs": 0,
    "tag": "Assets",
    "uom": "USD",
    "value": 5608533000,
    "version": "us-gaap/2009",
  },
  Object {
    "adsh": "0001193125-09-165391",
    "coreg": "",
    "ddate": 20090630,
    "footnote": "",
    "qtrs": 0,
    "tag": "LiabilitiesAndStockholdersEquity",
    "uom": "USD",
    "value": 24487332000,
    "version": "us-gaap/2008",
  },
  Object {
    "adsh": "0001144204-09-046684",
    "coreg": "",
    "ddate": 20090630,
    "footnote": "",
    "qtrs": 2,
    "tag": "OperatingIncomeLoss",
    "uom": "USD",
    "value": 257111000,
    "version": "us-gaap/2009",
  },
  Object {
    "adsh": "0001193125-09-165642",
    "coreg": "",
    "ddate": 20090630,
    "footnote": "",
    "qtrs": 1,
    "tag": "NetInvestmentIncome",
    "uom": "USD",
    "value": 174932000,
    "version": "us-gaap/2008",
  },
  Object {
    "adsh": "0001193125-09-160797",
    "coreg": "",
    "ddate": 20080630,
    "footnote": "",
    "qtrs": 2,
    "tag": "IncreaseDecreaseInAccountsReceivable",
    "uom": "USD",
    "value": -152000000,
    "version": "us-gaap/2008",
  },
  Object {
    "adsh": "0000950123-09-042090",
    "coreg": "",
    "ddate": 20080731,
    "footnote": "",
    "qtrs": 0,
    "tag": "SeniorLongTermNotes",
    "uom": "USD",
    "value": 545220000,
    "version": "us-gaap/2009",
  },
  Object {
    "adsh": "0001193125-09-166917",
    "coreg": "",
    "ddate": 20090630,
    "footnote": "",
    "qtrs": 2,
    "tag": "DepreciationAndAmortization",
    "uom": "USD",
    "value": 265300000,
    "version": "us-gaap/2008",
  },
  Object {
    "adsh": "0001193125-09-169449",
    "coreg": "",
    "ddate": 20080630,
    "footnote": "",
    "qtrs": 2,
    "tag": "AdjustmentsToAdditionalPaidInCapitalTaxEffectFromShareBasedCompensation",
    "uom": "USD",
    "value": 50000000,
    "version": "us-gaap/2008",
  },
  Object {
    "adsh": "0001193125-09-154417",
    "coreg": "",
    "ddate": 20090630,
    "footnote": "",
    "qtrs": 0,
    "tag": "Assets",
    "uom": "USD",
    "value": 56545000000,
    "version": "us-gaap/2009",
  },
  Object {
    "adsh": "0001193125-09-163743",
    "coreg": "",
    "ddate": 20080630,
    "footnote": "",
    "qtrs": 1,
    "tag": "GasGatheringTransportationMarketingAndProcessingCosts",
    "uom": "USD",
    "value": 10944000,
    "version": "us-gaap/2008",
  },
  Object {
    "adsh": "0000950123-09-031199",
    "coreg": "",
    "ddate": 20080630,
    "footnote": "",
    "qtrs": 1,
    "tag": "RestructuringCharges",
    "uom": "USD",
    "value": 1000000,
    "version": "us-gaap/2009",
  },
  Object {
    "adsh": "0001104659-09-047410",
    "coreg": "",
    "ddate": 20090630,
    "footnote": "",
    "qtrs": 1,
    "tag": "IncomeTaxExpenseBenefit",
    "uom": "USD",
    "value": 33900000,
    "version": "us-gaap/2009",
  },
  Object {
    "adsh": "0000950123-09-032022",
    "coreg": "ParentCompany",
    "ddate": 20090630,
    "footnote": "",
    "qtrs": 0,
    "tag": "InsuranceReceivables",
    "uom": "USD",
    "value": 39214000,
    "version": "0000950123-09-032022",
  },
  Object {
    "adsh": "0001193125-09-170830",
    "coreg": "",
    "ddate": 20081231,
    "footnote": "",
    "qtrs": 0,
    "tag": "OtherAssetsCurrent",
    "uom": "USD",
    "value": 344000000,
    "version": "us-gaap/2009",
  },
  Object {
    "adsh": "0000818479-09-000123",
    "coreg": "",
    "ddate": 20090630,
    "footnote": "",
    "qtrs": 2,
    "tag": "OperatingIncomeLoss",
    "uom": "USD",
    "value": 184879000,
    "version": "us-gaap/2009",
  },
  Object {
    "adsh": "0000950123-09-030860",
    "coreg": "",
    "ddate": 20090630,
    "footnote": "",
    "qtrs": 2,
    "tag": "OtherNoncashExpense",
    "uom": "USD",
    "value": 421000000,
    "version": "us-gaap/2009",
  },
  Object {
    "adsh": "0001047469-09-007134",
    "coreg": "",
    "ddate": 20081231,
    "footnote": "",
    "qtrs": 0,
    "tag": "StockholdersEquityIncludingPortionAttributableToNoncontrollingInterest",
    "uom": "USD",
    "value": 6039400000,
    "version": "us-gaap/2009",
  },
  Object {
    "adsh": "0000950123-09-029387",
    "coreg": "",
    "ddate": 20080630,
    "footnote": "",
    "qtrs": 1,
    "tag": "WeightedAverageNumberOfDilutedSharesOutstanding",
    "uom": "shares",
    "value": 157818000,
    "version": "us-gaap/2009",
  },
  Object {
    "adsh": "0001193125-09-169479",
    "coreg": "",
    "ddate": 20081231,
    "footnote": "",
    "qtrs": 0,
    "tag": "TimeDeposits",
    "uom": "USD",
    "value": 7667167000,
    "version": "us-gaap/2008",
  },
  Object {
    "adsh": "0001193125-09-168802",
    "coreg": "",
    "ddate": 20090630,
    "footnote": "",
    "qtrs": 0,
    "tag": "PropertyPlantAndEquipmentNet",
    "uom": "USD",
    "value": 8297448000,
    "version": "us-gaap/2008",
  },
  Object {
    "adsh": "0001193125-09-170393",
    "coreg": "",
    "ddate": 20080630,
    "footnote": "",
    "qtrs": 2,
    "tag": "BusinessCombinationAcquisitionRelatedCosts",
    "uom": "USD",
    "value": 36500000,
    "version": "0001193125-09-170393",
  },
  Object {
    "adsh": "0001193125-09-165464",
    "coreg": "",
    "ddate": 20090630,
    "footnote": "",
    "qtrs": 1,
    "tag": "IncomeTaxExpenseBenefit",
    "uom": "USD",
    "value": -395000,
    "version": "us-gaap/2008",
  },
  Object {
    "adsh": "0000950123-09-032642",
    "coreg": "",
    "ddate": 20090630,
    "footnote": "",
    "qtrs": 1,
    "tag": "InterestAndDividendIncomeOperating",
    "uom": "USD",
    "value": 994000000,
    "version": "us-gaap/2009",
  },
  Object {
    "adsh": "0001167379-09-000099",
    "coreg": "",
    "ddate": 20080630,
    "footnote": "",
    "qtrs": 2,
    "tag": "IncomeTaxExpenseBenefit",
    "uom": "USD",
    "value": 157000000,
    "version": "us-gaap/2009",
  },
  Object {
    "adsh": "0000040545-09-000049",
    "coreg": "Subsidiaries",
    "ddate": 20090630,
    "footnote": "",
    "qtrs": 1,
    "tag": "IncomeLossFromDiscontinuedOperationsNetOfTax",
    "uom": "USD",
    "value": -194000000,
    "version": "us-gaap/2009",
  },
  Object {
    "adsh": "0000930413-09-004048",
    "coreg": "",
    "ddate": 20090630,
    "footnote": "",
    "qtrs": 1,
    "tag": "MinorityInterestInNetIncomeLossPreferredUnitHolders",
    "uom": "USD",
    "value": 4571000,
    "version": "us-gaap/2009",
  },
  Object {
    "adsh": "0001193125-09-191566",
    "coreg": "",
    "ddate": 20081231,
    "footnote": "",
    "qtrs": 4,
    "tag": "IncreaseDecreaseInAccountsReceivable",
    "uom": "USD",
    "value": 451000000,
    "version": "us-gaap/2009",
  },
  Object {
    "adsh": "0001193125-09-170393",
    "coreg": "",
    "ddate": 20080630,
    "footnote": "",
    "qtrs": 2,
    "tag": "EarningsPerShareDiluted",
    "uom": "USD",
    "value": 0.18,
    "version": "us-gaap/2008",
  },
  Object {
    "adsh": "0001047469-09-007384",
    "coreg": "",
    "ddate": 20090630,
    "footnote": "",
    "qtrs": 0,
    "tag": "Goodwill",
    "uom": "USD",
    "value": 4600000,
    "version": "us-gaap/2009",
  },
  Object {
    "adsh": "0001193125-09-183869",
    "coreg": "",
    "ddate": 20080630,
    "footnote": "",
    "qtrs": 2,
    "tag": "ShareBasedCompensation",
    "uom": "USD",
    "value": 12040000,
    "version": "us-gaap/2009",
  },
  Object {
    "adsh": "0001047469-09-008166",
    "coreg": "",
    "ddate": 20090731,
    "footnote": "",
    "qtrs": 1,
    "tag": "CommonStockDividendsPerShareDeclared",
    "uom": "USD",
    "value": 0.16,
    "version": "us-gaap/2009",
  },
  Object {
    "adsh": "0000086521-09-000047",
    "coreg": "ParentCompany",
    "ddate": 20081231,
    "footnote": "",
    "qtrs": 0,
    "tag": "RegulatoryLiabilitiesArisingFromRemovalObligations",
    "uom": "USD",
    "value": 2430000000,
    "version": "0000086521-09-000047",
  },
  Object {
    "adsh": "0000950123-09-032827",
    "coreg": "",
    "ddate": 20090630,
    "footnote": "",
    "qtrs": 0,
    "tag": "CommonStockValue",
    "uom": "USD",
    "value": 2000,
    "version": "us-gaap/2009",
  },
  Object {
    "adsh": "0001193125-09-167155",
    "coreg": "",
    "ddate": 20080630,
    "footnote": "",
    "qtrs": 2,
    "tag": "PaymentsForRepurchaseOfCommonStock",
    "uom": "USD",
    "value": 296566000,
    "version": "us-gaap/2008",
  },
  Object {
    "adsh": "0001193125-09-158007",
    "coreg": "",
    "ddate": 20081231,
    "footnote": "",
    "qtrs": 0,
    "tag": "MinorityInterest",
    "uom": "USD",
    "value": 72000000,
    "version": "us-gaap/2008",
  },
  Object {
    "adsh": "0001157523-09-005774",
    "coreg": "",
    "ddate": 20081231,
    "footnote": "",
    "qtrs": 0,
    "tag": "LiabilitiesAndStockholdersEquity",
    "uom": "USD",
    "value": 111148000000,
    "version": "us-gaap/2009",
  },
  Object {
    "adsh": "0001193125-09-168935",
    "coreg": "",
    "ddate": 20090630,
    "footnote": "",
    "qtrs": 1,
    "tag": "InterestExpenseDeposits",
    "uom": "USD",
    "value": 2082000000,
    "version": "us-gaap/2008",
  },
  Object {
    "adsh": "0001310243-09-000085",
    "coreg": "",
    "ddate": 20090630,
    "footnote": "",
    "qtrs": 2,
    "tag": "CashAndCashEquivalentsPeriodIncreaseDecrease",
    "uom": "USD",
    "value": -8822000,
    "version": "us-gaap/2008",
  },
  Object {
    "adsh": "0001104659-09-053270",
    "coreg": "",
    "ddate": 20090630,
    "footnote": "",
    "qtrs": 2,
    "tag": "ForeignCurrencyTransactionGainLossUnrealized",
    "uom": "USD",
    "value": -413000000,
    "version": "us-gaap/2009",
  },
  Object {
    "adsh": "0001310243-09-000085",
    "coreg": "",
    "ddate": 20080630,
    "footnote": "",
    "qtrs": 2,
    "tag": "RepaymentsOfNotesPayable",
    "uom": "USD",
    "value": 12485000,
    "version": "us-gaap/2008",
  },
  Object {
    "adsh": "0001193125-09-159615",
    "coreg": "",
    "ddate": 20090630,
    "footnote": "",
    "qtrs": 3,
    "tag": "IncreaseDecreaseInAccruedIncomeTaxesPayable",
    "uom": "USD",
    "value": 105000000,
    "version": "us-gaap/2008",
  },
  Object {
    "adsh": "0001193125-09-162276",
    "coreg": "",
    "ddate": 20081231,
    "footnote": "",
    "qtrs": 0,
    "tag": "LongTermDebt",
    "uom": "USD",
    "value": 3293400000,
    "version": "us-gaap/2008",
  },
  Object {
    "adsh": "0000950123-09-032414",
    "coreg": "",
    "ddate": 20090630,
    "footnote": "",
    "qtrs": 0,
    "tag": "TreasuryStockValue",
    "uom": "USD",
    "value": 281000000,
    "version": "us-gaap/2009",
  },
  Object {
    "adsh": "0001047469-09-006994",
    "coreg": "",
    "ddate": 20081231,
    "footnote": "",
    "qtrs": 0,
    "tag": "LiabilitiesAndStockholdersEquity",
    "uom": "USD",
    "value": 20141000000,
    "version": "us-gaap/2009",
  },
  Object {
    "adsh": "0001193125-09-155929",
    "coreg": "",
    "ddate": 20090630,
    "footnote": "",
    "qtrs": 1,
    "tag": "ResearchAndDevelopmentExpense",
    "uom": "USD",
    "value": 41700000,
    "version": "us-gaap/2008",
  },
  Object {
    "adsh": "0001193125-09-172858",
    "coreg": "",
    "ddate": 20081231,
    "footnote": "",
    "qtrs": 0,
    "tag": "DerivativeLiabilitiesCurrent",
    "uom": "USD",
    "value": 49561000,
    "version": "us-gaap/2008",
  },
  Object {
    "adsh": "0000018230-09-000331",
    "coreg": "",
    "ddate": 20080630,
    "footnote": "",
    "qtrs": 2,
    "tag": "StockRepurchaseDerivativeContract",
    "uom": "USD",
    "value": -10000000,
    "version": "0000018230-09-000331",
  },
  Object {
    "adsh": "0001193125-09-163299",
    "coreg": "",
    "ddate": 20090630,
    "footnote": "",
    "qtrs": 0,
    "tag": "DeferredTaxLiabilitiesNoncurrent",
    "uom": "USD",
    "value": 9857000000,
    "version": "us-gaap/2008",
  },
  Object {
    "adsh": "0000277135-09-000023",
    "coreg": "",
    "ddate": 20080630,
    "footnote": "",
    "qtrs": 1,
    "tag": "GrossProfit",
    "uom": "USD",
    "value": 705877000,
    "version": "us-gaap/2008",
  },
  Object {
    "adsh": "0000950123-09-038339",
    "coreg": "",
    "ddate": 20090630,
    "footnote": "",
    "qtrs": 1,
    "tag": "OperatingIncomeLoss",
    "uom": "USD",
    "value": 351000000,
    "version": "us-gaap/2009",
  },
]
`;
